### PR TITLE
Native mid-circuit measurements support composite measurements, etc. (phase 2)

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
This PR is a follow-up on #5088 .

[Epic 50811](https://app.shortcut.com/xanaduai/epic/50811)  states:

- [x] The deferred measurement principle is not used
- [x] Each shot is a new exploration of the tree of possible mid-circuit measurement results
- [x] The mid-circuit measurement results are correctly exposed to PennyLane's frontend
- [x] Compatible with qubit reuse
- [x] Compatible with qubit reset
- [x] Compatible with qubit postselection
- [x] Compatible with existing support for conditioning in PennyLane
  - [x] Addition, subtraction and multiplication of one or more mid-circuit measurement results

The current PR tackles the requirements above as indicated by the task list.

**Description of the Change:**

- `apply_mid_measure` handles post-selection returning a zero state and sample when the state isn't properly normalized (which happens when the state cannot be normalized).
- The `concretize` method is added to MeasurementValue object to obtain a concrete value from a dictionary of measurement values. This allows handling composite measurements. Other modules like `simulate` are simplified accordingly. 
- Post-selected-zero-state samples are thrown away at the `simulate` level, handling this exception. This means the final number of samples is lower or equal to the total number of shots.
- Tests for all introduced features.

**Benefits:**
Mid-circuit measurements are natively supported.

**Possible Drawbacks:**
Slower in several cases (e.g. when few MCMs).

**Missing features:**
To do in follow-up PRs:

- [ ] parameter-shift support
- [ ] optimisation

**Related GitHub Issues:**
[sc-54987]
[sc-55666]
[sc-55667]
[sc-55668]
[sc-55922]

### CODE REVIEW

**Comment by mudit2812** [comment]
Lots of merge conflicts so the diff is really messed up. Will take a look after fixed.

**Comment by vincentmr** [comment]
> Lots of merge conflicts so the diff is really messed up. Will take a look after fixed.

Indeed, I got a lot of feedback from @timmysilv and @albi3ro on #5088 on Friday and the code changed a lot. Once #5088 is merged, I'll refactor this one and open it for review.

**Comment by mudit2812** [comment]
> > Lots of merge conflicts so the diff is really messed up. Will take a look after fixed.
> 
> Indeed, I got a lot of feedback from @timmysilv and @albi3ro on #5088 on Friday and the code changed a lot. Once #5088 is merged, I'll refactor this one and open it for review.

Thanks @vincentmr . Could you convert this to a draft in the mean time?

File: .github/workflows/core_tests_durations.json
Comment (mudit2812) [inline .github/workflows/core_tests_durations.json:1]: Is this file related to this PR?

Comment (vincentmr) [reply .github/workflows/core_tests_durations.json:1]: Sort of. The core tests were getting quite long and poorly distributed by pytest-xdist. I thus timed the core tests and pytest now uses this file to split the tests. You'll see some runners have 100-200 tests and one of them has ~20k, and hence the need to split according to duration.

Comment (mudit2812) [reply .github/workflows/core_tests_durations.json:1]: Ah gotcha, maybe worth getting input from other people on the team as well. This shouldn't affect CI if it's only for pytest-xdist.

File: .github/workflows/interface-unit-tests.yml
Comment (mudit2812) [inline .github/workflows/interface-unit-tests.yml:99]: Is this the number of runners per test type?

Comment (vincentmr) [reply .github/workflows/interface-unit-tests.yml:99]: Yes. I think core tests now take around a 100 minutes ==> split into 5.

Comment (timmysilv) [reply .github/workflows/interface-unit-tests.yml:99]: will have to chat with @rashidnhm, I think we're on a bit of a budget with workers per PR

File: pennylane/devices/qubit/apply_operation.py
Comment (timmysilv) [inline pennylane/devices/qubit/apply_operation.py:263]: why does this happen here, at the beginning of `apply_mid_measure`? I would have expected a check like this after some measurements are applied

File: pennylane/devices/qubit/simulate.py
Comment (timmysilv) [inline pennylane/devices/qubit/simulate.py:404]:
```suggestion
            raise TypeError(f"Unsupported measurement of {type(m).__name__}.")
```

File: pennylane/devices/qubit/simulate.py
Comment (timmysilv) [inline pennylane/devices/qubit/simulate.py:454]:
```suggestion
                f"Native mid-circuit measurement mode does not support {type(m).__name__} measurements."
```

File: pennylane/devices/qubit/simulate.py
Comment (timmysilv) [inline pennylane/devices/qubit/simulate.py:486]:
```suggestion
            f"Native mid-circuit measurement mode does not support {type(circuit_measurement).__name__} measurements."
```

File: pennylane/devices/qubit/simulate.py
Comment (timmysilv) [inline pennylane/devices/qubit/simulate.py:518]: can you explain what's happening here? I'm a little worried when we arbitrarily use a wire like `0`, and I am unsure of what happens if you make something like `ExpectationMP(wires=Wires(0))`. It technically creates an object, but it doesn't feel like it should be valid.

File: pennylane/measurements/mid_measure.py
Comment (timmysilv) [inline pennylane/measurements/mid_measure.py:403]: i love this, thank you for adding it

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py
Comment (timmysilv) [inline tests/devices/default_qubit/test_default_qubit_native_mcm.py:38]: for style points :p
```suggestion
    for key1, val1 in results1.items():
```

**Review by timmysilv — COMMENTED** [review]
looks awesome! just one real question

File: .github/workflows/interface-unit-tests.yml
Comment (vincentmr) [reply .github/workflows/interface-unit-tests.yml:99]: Increasing workers from ~30 to ~35 doesn't seem to bad compared with Lightning ~80 workers, but yeah let's ask @rashidnhm to weigh in.

File: pennylane/devices/qubit/simulate.py
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:518]: When we have a composite measurement, say `qml.count((op=(m0 - 2 * m1) * m2 + 7))` where `m0`, `m1`, `m2` are MCMs on wires 0, 1 and 0 respectively, `mv.wires` will be `<Wires = [0, 1]>`. But `mcm_samples = np.array([mv.concretize(dct) for dct in samples]).reshape((-1, 1))` is not a 2-wire sample, it is like a 1-wire sample sampling from the interval `[5, 8]`. I thus trick the code by introducing `wires = qml.wires.Wires(0)` so I can use `process_samples` down the line.

File: doc/releases/changelog-dev.md
Comment (mudit2812) [inline doc/releases/changelog-dev.md:175]: I think these can be merged into a single entry.

Comment (mudit2812) [inline doc/releases/changelog-dev.md:175]: Let me know what you think about this. I think either is fine though.

```suggestion
* The `default.qubit` device now supports mid-circuit measurements without the use of the deferred measurement
  principle when using finite shots. This includes all features supported with `qml.defer_measurements`, including
  classical control, collecting statistics, and postselection.
  [(#5120)](https://github.com/PennyLaneAI/pennylane/pull/5120)
```

Comment (vincentmr) [reply doc/releases/changelog-dev.md]: Thanks @mudit2812 , read much better for a user. I copy-pasted and added a bit to it.

File: pennylane/devices/qubit/apply_operation.py
Comment (timmysilv) [reply pennylane/devices/qubit/apply_operation.py:263]: but because this is at the beginning of the function, doesn't that mean we'll only avoid that issue if someone postselects and makes the state disappear, and then they do _another_ `qml.measure`? which means that if they don't do another, we won't get the benefit of exiting early

File: pennylane/devices/qubit/simulate.py
Comment (timmysilv) [reply pennylane/devices/qubit/simulate.py:518]: hmm.. this is definitely tricking the `SampleMeasurement` interface into doing something it isn't supposed to do. It works, but it makes me nervous. This seems like the interface isn't meeting our needs, so I won't block the PR on it but I wanna make sure people know what's happening.

File: pennylane/devices/qubit/apply_operation.py
Comment (vincentmr) [reply pennylane/devices/qubit/apply_operation.py:263]: You are right, this is a very minor optimization since all other operations still get applied until the end of the tape. 

In the deferred measurement implementation, I believe non-normalizable states are just forcefully normalized (to NaNs) and then the rest of the circuit is applied on NaNs, yielding NaNs which are discarded at the end. If we're ok with that, I would return a NaN state instead of zeros and exit from `get_final_state` early as soon as a NaN is detected in `state`. How about that?

File: pennylane/devices/qubit/simulate.py
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:518]: Let me add the explanation above in a comment then. I could also try to hijack `process_samples` to use the observable branch of `_samples_to_counts` instead of the `wires` branch. Would that be better, or just more of the same?

Comment (timmysilv) [reply pennylane/devices/qubit/simulate.py:518]: imo more of the same. a comment would be perfect for now 😄

File: .github/workflows/interface-unit-tests.yml
Comment (rashidnhm) [reply .github/workflows/interface-unit-tests.yml:99]: It's not about the total number of workers you use but how many you use *at once*. If we save a lot of time bumping this then it might be worth the trade off.


Also ... I see `tf-tests: 3` was removed? Are we okay with tf-tests only using 1 worker at a time?

Comment (vincentmr) [reply .github/workflows/interface-unit-tests.yml:99]: It moved down when I sorted the fields but it should still be there.

Comment (timmysilv) [reply .github/workflows/interface-unit-tests.yml:99]: then this sounds good to me!

File: pennylane/devices/qubit/apply_operation.py
Comment (mudit2812) [inline pennylane/devices/qubit/apply_operation.py:278]: If we don't care about this dict when postselection doesn't match the sample, there's not really a need to update it here or above.

File: pennylane/devices/qubit/apply_operation.py
Comment (mudit2812) [reply pennylane/devices/qubit/apply_operation.py:263]: I would actually be in favour of exiting `get_final_state` early when postselection doesn't match the sample, because that could have a massive speed up in circuits where the probability of the postselected state is low and there are many gates after postselection. @trbromley do you have thoughts here?

Comment (vincentmr) [reply pennylane/devices/qubit/apply_operation.py:263]: Trying to exit `get_final_state` prematurely breaks a bunch of tests because autodiff has trouble tracking the NaNs or the `norm` function. I'd suggest we just leave it at that for now.

File: tests/measurements/test_expval.py
Comment (mudit2812) [inline tests/measurements/test_expval.py:130]: I feel like rather than not testing `defer_measurements` with finite shots there should be a new test for `defer_measurements`. This is probably also true for other MPs and it would make sense for such tests to be in `tests/transforms/test_defer_measurements.py`. Out of scope for this PR. I've added a shortcut story.

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py
Comment (mudit2812) [inline tests/devices/default_qubit/test_default_qubit_native_mcm.py:60]: Could you add light docstrings for these helpers?

File: tests/devices/default_qubit/test_default_qubit_native_mcm.py
Comment (mudit2812) [inline tests/devices/default_qubit/test_default_qubit_native_mcm.py:115]:
```suggestion
def test_parse_native_mid_circuit_measurements_unsupported_meas(measurement):
```

**Review by mudit2812 — COMMENTED** [review]
Thanks Vincent! Everything looks mostly okay, I've left a few comments. I'll come back for another review.

File: pennylane/devices/qubit/apply_operation.py
Comment (vincentmr) [reply pennylane/devices/qubit/apply_operation.py:278]: If we do not exit early in the caller function, then we still need to assign a value so that posterior `Conditional`s can be apply without throwing.

File: tests/measurements/test_expval.py
Comment (vincentmr) [reply tests/measurements/test_expval.py:130]: I've added some MPs tests for `defer_measurements`.

File: pennylane/devices/qubit/simulate.py
Comment (albi3ro) [inline pennylane/devices/qubit/simulate.py:307]: What would give rise to `one_shot_meas` being `None`?

File: pennylane/devices/qubit/simulate.py
Comment (albi3ro) [inline pennylane/devices/qubit/simulate.py:456]:
```suggestion
        if m.mv:
```

I feel like I'm missing something. This should be the same, right?

File: pennylane/devices/qubit/simulate.py
Comment (albi3ro) [inline pennylane/devices/qubit/simulate.py:450]: I'm just curious about the difference between preallocating the array and then updating it versus initializing it to `[]` and appending to it.  If we are appending to a list instead of updating a list, we could get rid of the dependence on `i`.

File: pennylane/devices/qubit/simulate.py
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:307]: This comes out `None` when we post-select a zero-prob branch of the statevector. It is just a convention allowing to know when to discard the shot.

File: pennylane/devices/qubit/simulate.py
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py]: Yes, better.

File: pennylane/devices/qubit/simulate.py
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py]: Let's do it like that, looks more pythonic. We still need the dependence on `i` because we may need to pass `all_shot_meas[i]`. We can't zip it with `circuit.measurements` since it may be `None` in certain cases.

**Review by mudit2812 — APPROVED** [review]
Thanks @vincentmr ! This looks good to me. Very excited to get this in. I'm still a bit worried about the core test runners, but everyone seems okay with it so let's do it

File: .github/workflows/core_tests_durations.json
Comment (trbromley) [inline .github/workflows/core_tests_durations.json:1]: This 27589 line file scares me :fearful: Is this needed?

Comment (timmysilv) [reply .github/workflows/core_tests_durations.json:1]: discussed in [this conversation](https://github.com/PennyLaneAI/pennylane/pull/5120#discussion_r1481768689) above, it's an auto-generated file to `pytest-split` does a better job grouping core tests together.

Comment (vincentmr) [reply .github/workflows/core_tests_durations.json:1]: Thanks @timmysilv . Yes @trbromley , the TLDR is core tests are taking too long and we need time data to properly accelerate the tests via load distribution (otherwise load imbalance does away with virtually any benefit).

File: pennylane/devices/qubit/simulate.py
Comment (timmysilv) [inline pennylane/devices/qubit/simulate.py:304]: i suppose this isn't related to this PR strictly, but does this account for shot-vectors correctly? eg. if a user uses `shots=[1000, 2000]`, will they get 2 results (one from the first 1000 shots, discarding the invalid ones, and another from the final 2000 shots)?

**Review by timmysilv — COMMENTED** [review]
one last concern, then I'd say this is good to go!

File: pennylane/devices/qubit/simulate.py
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:304]: That's a good question. Yes, I thought that's what shot-vectors do, but I may be wrong there. I think what could happen is that a user will get a warning for all shots. Or maybe a warning for some items in the shot-vector and not others. Is that what you're wondering? Maybe I can add a test for that, if so.

Comment (timmysilv) [reply pennylane/devices/qubit/simulate.py:304]: never mind, ignore me! I was wondering where the nested for-loop is for the case of partitioned shots, but I scrolled up and I see that you factored it out into a nested call 3 lines above this one :p lgtm

**Review by timmysilv — APPROVED** [review]

**Pull request merged**
